### ISSUE TITLE
[nat] Packets not translated when using 3 cores

### ISSUE THREAD

**Issue by tbarbette** [ndsl-kaist/mOS-networking-stack#17]
Hi,

The midstat application works great with any number of cores (I tried 1 to 4), with an expected more or less linear improvement, so it works.
However, the NAT application does not translate some SYN+ACK when using 3 cores, and only with 3 cores. It is as if the packet did go through untouched.
Any idea where that could come from? Did I miss something?

Thanks,
Tom

**Comment by a-jamshed** [comment]
@tbarbette,

I am not sure about the exact reason behind this anomaly, but I can make a guess. The mOS stack is probably running out of memory (out of flow records???) which means that some packets do not trigger event callbacks. The default behavior (I think) for such case is to forward the packet without analyzing it. Can you try increasing ``max_concurrency`` value from ``mos.conf`` file?

I remember that one of my co-author(s) worked on a new version of ``nat`` recently. Let me ask them about its status. Also, they may also be able to give more ideas regarding the reported issue as well.

**Comment by tbarbette** [comment]
Thanks for the help ! I doubt it's the issue at hand here. I've tried a few flows by hand after finding the issue. max_concurrency is hardly a problem here. Moreover, the test works with 2/4 cores and with 3 with midstat.

**Comment by archon159** [comment]
@tbarbette,

Hello, I'm Sean Lee now working in mOS team.

I'm sorry about replying to your comment late. 

I am testing mOS applications according to your report.

While testing, I could found an issue that NAT does not translate some packets.

The issue have been reproduced regardless of cores but it is related to max concurrency, especially more than 200.

Nevertheless, this issue seems orthogonal to your issue as your issue is related to the number of cores.

I used linux client (Apache Benchmark) and linux server (Nginx) for the experiment.

Each machine uses Intel X520 NICs.

Would you please inform your detailed setup or configuration used in your experiment?

It may contain following details.
- Benchmark application or code (client/server) used in the experiment.
- NIC information
- Machine information
- Whether there were only SYN+ACK packets that were not translated.
- max_concurrency/cores
- How did you know that midstat works well?
- I/O module that you use in mOS (DPDK, PSIO...)
- tcp timeout/timewait interval that you set in mOS configuration
- netdev that you set in mOS configuration


Thanks,
Sean

**Comment by tbarbette** [comment]
Hi !

- WRK, but it is similar to AB. With wget, same problem.
- Intel XL710 2*40G  adapters
- Xeon E5 v3 and v4, nothing of importance here
- Not sure, and the setup is not online anymore. From what I wrote, I would say the SYN were translated.
- See below important parts of the config :     
```
[...]
        forward = 1

	# devices used for MOS applications [mandatory]
	netdev {
		dpdk0 0x000f    #when using 4 cores, (I use 1 for 1, 3 for 2, 7 for 3, ...)
		dpdk1 0x000f    #idem
	}

	nic_forward_table {	
        dpdk0 dpdk1
	}
[...]
       max_concurrency = 100000

	# disable the ring buffer [optional / default : 0]
	# use disabled buffered managment only for standalone monitors.
	# end host applications always need recv buffers for TCP!
	# no_ring_buffers = 1

	# receive buffer size of sockets [optional / default : 8192]
	# rmem_size = 8192

	# send buffer size of sockets [optional / default : 8192]
	# wmem_size = 8192

	# tcp timewait seconds [optional / default : 0]
	tcp_tw_interval = 2

	# tcp timeout seconds [optional / default : 30]
	# (set tcp_timeout = -1 to disable timeout checking)
	tcp_timeout = 30

```
- Packets pass through (indeed, if they are passed untouched, I would not know about that...)
- DPDK
- See above
- See above

Thanks,
Tom

**Comment by archon159** [comment]
Hello, Tom

Thank you for reply.
It would be helpful to reproduce the issue you've told.

Thanks,
Sean

**Comment by archon159** [comment]
Hello Tom

First, I would like to inform you that the issues you reported are not reproduced in the route table environment that I am currently experimenting with.
There is the only difference between your setup and my setup, whether using the nic forward table or the route table.
As a result, I was preparing for the experiment using the nic forward table but it is delayed due to a hardware issue.
I am sorry for the delay.
If you want to use mOS as soon as possible, I recommend using route table.
Apart from that, I will do my best to resolve the issue.

Thanks,
Sean

**Comment by archon159** [comment]
@tbarbette 
Hello Tom

I could reproduce the issue you've mentioned.
The issue did not occur with intel X520 10G NIC.
However, XL710 40G NIC reproduced the issue which 
NAT built on mOS with 3 cores does not translate SYN-ACK packets.
(With both setup, nic forwarding table and routing table. Sorry for the previous comment.)
I expect that this issue is related to RSS-Key, but more details need debugging.
Thank you for reporting this issue.
I'll let you know as soon as the issue is solved.

Thanks,
Sean

**Comment by tbarbette** [comment]
Thanks for your concern anyway ! Great to hear the problem is being solved.

**Comment by ygmoon** [comment]
@tbarbette 

Hi Tom,

Thanks to your report, we've corrected the RSS mechanism in d88f3b1df1fbe5f88a4201c9af0205ca2006fd30. I leave some details here.

> 1. Problem description
> - When mOS NAT translate a packet, GetRSSCPUCore() is called to calculate the expected Rx queue number for receiving the "returning" packets for the given 4-tuple. 
> - However, when it runs on N cores (where N is not 2 ^ n) with XL710 NIC, the result of GetRSSCPUCore() is different from the actual Rx queue number who gets the packet.
> 
> 2. Does the mismatch come from RSS hash value?
> - No. Although X520 NIC (Intel 10GbE) and XL710 NIC (Intel 40GbE) use different Random Secret Key (RSK) for RSS (40-bytes vs. 52-bytes), we can guarantee the same RSS hash result if the 52-byte RSK is generated from the 40-byte RSK + 12-byte zero-padding.
> - When I compare the RSS hash value calculated by GetRSSHash() in mTCP against the hash value calculated by NIC (delivered together via rte_mbuf->hash.rss field), they are always the same.
> 
> 3. Then why does the mismatch happen?
> - In short, it comes from the mismatch of the function from the RSS hash value to the Rx queue number. More specifically, RSS mechanisms for each NIC are as follows: 
> - X520 NIC:  (Rx queue number) = (7 LS bits of RSS hash value) mod N (See Section 7.1.2.8 of https://www.intel.com/content/dam/www/public/us/en/documents/datasheets/82599-10-gbe-controller-datasheet.pdf)
> - XL710 NIC: (Rx queue number) = (9 LS bits of RSS hash value) mod N (See Section 7.1.8 of https://www.intel.com/content/dam/www/public/us/en/documents/datasheets/xl710-10-40-controller-datasheet.pdf)
> - GetRSSCPUCore() of mOS has been following the RSS mechanism of X520 NIC, and it was not working correctly with the XL710 NIC.
> 
> 4. Why mOS NAT has been working fine with 2 ^ n CPU cores (1, 2, 4, 8, ..)?
> - When the divisor is 2 ^ n, modulo operation is just same as taking n LS bits. In other words, if the number of Rx queues is 2 ^ n,
> - X520 NIC:  (RX queue number) = n LS bits of (7 LS bits of RSS hash value)
> - XL710 NIC: (RX queue number) = n LS bits of (9 LS bits of RSS hash value) 
> - Therefore, as far as n <= 7 (the number of Rx queues <= 128), the results would be the same.
> 
> 5. Proposed solution and results
> - We can simply detect the driver name (either ixgbe or i40e), and set the correct value RSS_BIT_MASK for each driver.
> - I have tested mOS NAT with up to 32 cores with XL710, and up to 8 cores with X520 NIC. It seems working fine with any number of CPU cores.

**Comment by tbarbette** [comment]
Great !
